enclave/crypto/openssl/crl.c

```diff
@@ -0,0 +1,277 @@
+{
+    oe_result_t result = OE_CRYPTO_ERROR;
+    crl_t* impl = (crl_t*)crl;
+    BIO* bio = NULL;
+    X509_CRL* x509_crl = NULL;
+
+    /* Clear the implementation */
+    if (impl)
+        memset(impl, 0, sizeof(crl_t));
+
+    /* Check for invalid parameters */
+    if (!der_data || !der_size || der_size > OE_INT_MAX || !crl)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    /* Create a BIO for reading the DER-formatted data */
+    if (!(bio = BIO_new_mem_buf(der_data, (int)der_size)))
+        goto done;
+
+    /* Read BIO into X509_CRL object */
+    if (!(x509_crl = d2i_X509_CRL_bio(bio, NULL)))
+        goto done;
+
+    /* Initialize the implementation */
+    _crl_init(impl, x509_crl);
+    x509_crl = NULL;
+
+    result = OE_OK;
+
+done:
+
+    if (x509_crl)
+        X509_CRL_free(x509_crl);
+
+    if (bio)
+        BIO_free(bio);
+
+    return result;
+}
+
+oe_result_t oe_crl_read_pem(
+    oe_crl_t* crl,
+    const uint8_t* pem_data,
+    size_t pem_size)
+{
+    oe_result_t result = OE_CRYPTO_ERROR;
+    crl_t* impl = (crl_t*)crl;
+    BIO* bio = NULL;
+    X509_CRL* x509_crl = NULL;
+
+    /* Clear the implementation */
+    if (impl)
+        memset(impl, 0, sizeof(crl_t));
+
+    /* Check for invalid parameters */
+    if (!pem_data || !pem_size || pem_size > OE_INT_MAX || !crl)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    /* Create a BIO for reading the DER-formatted data */
+    if (!(bio = BIO_new_mem_buf(pem_data, (int)pem_size)))
+        goto done;
+
+    /* Read BIO into X509_CRL object */
+    if (!(x509_crl = PEM_read_bio_X509_CRL(bio, NULL, NULL, NULL)))
+        goto done;
+
+    /* Initialize the implementation */
+    _crl_init(impl, x509_crl);
+    x509_crl = NULL;
+
+    result = OE_OK;
+
+done:
+    if (x509_crl)
+        X509_CRL_free(x509_crl);
+
+    if (bio)
+        BIO_free(bio);
+
+    return result;
+}
+
+oe_result_t oe_crl_free(oe_crl_t* crl)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    crl_t* impl = (crl_t*)crl;
+
+    /* Check the parameter */
+    if (!crl_is_valid(impl))
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    /* Free the CRL */
+    _crl_free(impl);
+
+    result = OE_OK;
+
+done:
+    return result;
+}
+
+// Parse a string into a oe_datetime_t: example: "May 30 10:23:42 2018 GMT".
+// This format is specific to OpenSSL: produced by ASN1_TIME_print().
+static oe_result_t _string_to_date(const char* str, oe_datetime_t* date)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    char month[4];
+
+    memset(date, 0, sizeof(oe_datetime_t));
+
+    /* Convert the string to oe_datetime_t struct */
+    if (sscanf(
+            str,
+            "%3s %02u %02u:%02u:%02u %04u",
+            month,
+            &date->day,
+            &date->hours,
+            &date->minutes,
+            &date->seconds,
+            &date->year) != 6)
+    {
+        OE_RAISE(OE_FAILURE);
+    }
+
+    /* Convert the month string to integer */
+    {
+        static const char* _month[] = {"Jan",
```

Comment:
Again, can we just call strptime and not have to reimplement date/string conversion ourselves?